### DISCUSSION TITLE
Workspace Example

### DISCUSSION THREAD

**Discussion by cdietrich** [TypeFox/monaco-languageclient#783] [Q&A]
Hi,
do we meanwhile have an example for the following scenario

- ~~websocket~~web worker based language server
- client (react) maintains a list of (companion) files / workspace, server has no access to a physical filesystem
- a primary file is edited
- ideally the server does not need to support a special sidechannel to talk / be fed by the client

**Comment by kaisalmen** [comment]
@cdietrich we have some of the things. The [clangd example](https://github.com/TypeFox/monaco-languageclient#cpp--clangd-location) has language server running in web worker with a memory file system that drives the client file system via a message channel (driver -> follower). So, it is the other way around.

I don't understand what you mean with websocket based language server. That's what has been here all the time or do you mean something different?

A side channel for file sync seems to be the easier solution. What Is started for clangd is not finished and we only have a MessageChannel implementation for the main<->worker exchange. Doing the same via socket.io for a remote language servefr should be straightforward.

We are open to contributions or to new open specific enhancement issues.

**Reply by cdietrich** [reply]
the problem is that maintaining such a side channel is extra work and might need to be done by different teams depending on who owns the server and who owns the client.
we currently use the modelRef approach but as discussed previously this is discouraged

websocket is a typo. i meant web worker

**Comment by kaisalmen** [comment]
> the problem is that maintaining such a side channel is extra work and might need to be done by different teams depending on who owns the server and who owns the client.

Yes, I understand. Putting it on top of the LSP was the first idea, but it is problematic when you don't have control over the language server implementation. But when you have an abstract definition, there could also be a LSP based implementation, too.

**Reply by cdietrich** [reply]
the abstract lsp concept is "didOpen" and i assume this is what "modelRef" does under the hood

**Reply by kaisalmen** [reply]
Problem is that some language servers like pyright or Eclipse JDT ignore some requests and you don't have control. They assume being on the same file system as the client and it takes care. What is true in a VSCode desktop application.
